### ISSUE TITLE
Jinja Templating Causing Destructive Reformat

### ISSUE THREAD

**Issue by joelossher** [sqlfluff/sqlfluff#5641]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We encountered a few queries where running `fix` resulted in part of the join condition being removed.

This appears to be caused by a problematic interaction between jinja templating and rules `ST09` and `LT02`

### Expected Behaviour

`fix` should never change the semantics of the query

### Observed Behaviour

Running fix on this SQL

```sql
SELECT table_a.*
FROM schema.table_a
  INNER JOIN schema.table_b
    ON table_b.col_a = table_a.col_a
      AND table_b.col_b = {{ snapshot_date() }}
```

results in

```sql
SELECT table_a.*
FROM schema.table_a
  INNER JOIN schema.table_b
    ON
      table_b.col_a
      AND table_b.col_b = {{ snapshot_date() }}
```

The issue goes away if:
* `{{ snapshot_date() }}` is replaced with the result of render `CAST('2022-09-23' AS DATE)`
* The `ST09` error is fixed by changing `table_b.col_a = table_a.col_a` to `table_a.col_a = table_b.col_a`

### How to reproduce

I've been running `sqlfluff fix --dialect redshift data-models/python/noom_data_models/transforms/example/example.sql`

### Dialect

Redshift

### Version

sqlfluff, version 2.3.5
Python 3.9.18

### Configuration
```
[sqlfluff]
templater = jinja
sql_file_exts = .sql
# L016 - Line Length (unnecessary given our macro logic)
# L029 - Keywords should not be used as identifies (too many legacy tables rely on this)
# L034 - Statement ordering complexity (not-applicable)
exclude_rules = L016,L029,L034
large_file_skip_byte_limit=50000

[sqlfluff:indentation]
indent_unit = space
tab_space_size = 2
indented_joins = true
indented_using_on = true
template_blocks_indent = false

[sqlfluff:rules]
allow_scalar = True
unquoted_identifiers_policy = all
hanging_indents = False

[sqlfluff:layout:type:comma]
line_position = trailing


[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.keywords]
# Inconsistent capitalisation of keywords
capitalisation_policy = upper

[capitalisation.identifiers]
# Inconsistent capitalisation of unquoted identifiers.
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = upper

[sqlfluff:rules:references.quoting]
ignore_words=time,date

[sqlfluff:templater:jinja:macros]
snapshot_date = {% macro snapshot_date(delta_days=0, format_as="DATE") %}{% if format_as == 'DATE' %}{{ "CAST('2022-09-23' AS DATE)" }}{% else %}{{ "CAST('2022-09-23T00-00' AS TIMESTAMP)" }}{% endif %}{% endmacro %}

[sqlfluff:templater:jinja]
load_macros_from_path = noom_data_models/templates/

[sqlfluff:templater:jinja:context]
# A list of all variables from noom-data-models/templates/*.jinja
data_interval_start=none
data_interval_end=none
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jmc-bbk** [comment]
In the version of the query with the macro `{{ snapshot_date() }}`, the `join_on_condition` parse tree doesn't include the  `= table_a.col_a` comparison.

Example:

```
[L:  4, P:  5]      |                        join_on_condition:
[L:  4, P:  5]      |                            keyword:                          'ON'
[L:  4, P:  7]      |                            [META] (implicit) indent:
[L:  4, P:  7]      |                            newline:                          '\n'
[L:  5, P:  1]      |                            whitespace:                       '      '
[L:  5, P:  7]      |                            expression:
[L:  5, P:  7]      |                                column_reference:
[L:  5, P:  7]      |                                    naked_identifier:         'table_b'
[L:  5, P: 14]      |                                    dot:                      '.'
[L:  5, P: 15]      |                                    naked_identifier:         'col_a'
[L:  5, P: 20]      |                                newline:                      '\n'
[L:  6, P:  1]      |                                whitespace:                   '      '
[L:  6, P:  7]      |                                binary_operator:              'AND'
[L:  6, P: 10]      |                                whitespace:                   ' '
[L:  6, P: 11]      |                                column_reference:
[L:  6, P: 11]      |                                    naked_identifier:         'table_b'
[L:  6, P: 18]      |                                    dot:                      '.'
[L:  6, P: 19]      |                                    naked_identifier:         'col_b'
[L:  6, P: 24]      |                                whitespace:                   ' '
[L:  6, P: 25]      |                                comparison_operator:
[L:  6, P: 25]      |                                    raw_comparison_operator:  '='
[L:  6, P: 26]      |                                whitespace:                   ' '
[L:  6, P: 27]      |                                function:
[L:  6, P: 27]      |                                    function_name:
[L:  6, P: 27]      |                                        function_name_identifier:  'CAST'
[L:  6, P: 27]      |                                    bracketed:
[L:  6, P: 27]      |                                        start_bracket:        '('
[L:  6, P: 27]      |                                        [META] indent:
[L:  6, P: 27]      |                                        expression:
[L:  6, P: 27]      |                                            quoted_literal:   "'2022-09-23'"
[L:  6, P: 27]      |                                        whitespace:           ' '
[L:  6, P: 27]      |                                        keyword:              'AS'
[L:  6, P: 27]      |                                        whitespace:           ' '
[L:  6, P: 27]      |                                        data_type:
[L:  6, P: 27]      |                                            datetime_type_identifier:
[L:  6, P: 27]      |                                                keyword:      'DATE'
[L:  6, P: 27]      |                                        [META] dedent:
[L:  6, P: 27]      |                                        end_bracket:          ')'
```

Compare this to the parse tree of the same query without the macro, and `CAST('2023-09-23' AS DATE)` passed instead.

```
[L:  4, P:  5]      |                        join_on_condition:
[L:  4, P:  5]      |                            keyword:                          'ON'
[L:  4, P:  7]      |                            [META] (implicit) indent:
[L:  4, P:  7]      |                            whitespace:                       ' '
[L:  4, P:  8]      |                            expression:
[L:  4, P:  8]      |                                column_reference:
[L:  4, P:  8]      |                                    naked_identifier:         'table_b'
[L:  4, P: 15]      |                                    dot:                      '.'
[L:  4, P: 16]      |                                    naked_identifier:         'col_a'
[L:  4, P: 21]      |                                whitespace:                   ' '
[L:  4, P: 22]      |                                comparison_operator:
[L:  4, P: 22]      |                                    raw_comparison_operator:  '='
[L:  4, P: 23]      |                                whitespace:                   ' '
[L:  4, P: 24]      |                                column_reference:
[L:  4, P: 24]      |                                    naked_identifier:         'table_a'
[L:  4, P: 31]      |                                    dot:                      '.'
[L:  4, P: 32]      |                                    naked_identifier:         'col_a'
[L:  4, P: 37]      |                                newline:                      '\n'
[L:  5, P:  1]      |                                whitespace:                   '      '
[L:  5, P:  7]      |                                binary_operator:              'AND'
[L:  5, P: 10]      |                                whitespace:                   ' '
[L:  5, P: 11]      |                                column_reference:
[L:  5, P: 11]      |                                    naked_identifier:         'table_b'
[L:  5, P: 18]      |                                    dot:                      '.'
[L:  5, P: 19]      |                                    naked_identifier:         'col_b'
[L:  5, P: 24]      |                                whitespace:                   ' '
[L:  5, P: 25]      |                                comparison_operator:
[L:  5, P: 25]      |                                    raw_comparison_operator:  '='
[L:  5, P: 26]      |                                whitespace:                   ' '
[L:  5, P: 27]      |                                function:
[L:  5, P: 27]      |                                    function_name:
[L:  5, P: 27]      |                                        function_name_identifier:  'CAST'
[L:  5, P: 31]      |                                    bracketed:
[L:  5, P: 31]      |                                        start_bracket:        '('
[L:  5, P: 32]      |                                        [META] indent:
[L:  5, P: 32]      |                                        expression:
[L:  5, P: 32]      |                                            quoted_literal:   "'2023-09-23'"
[L:  5, P: 44]      |                                        whitespace:           ' '
[L:  5, P: 45]      |                                        keyword:              'AS'
[L:  5, P: 47]      |                                        whitespace:           ' '
[L:  5, P: 48]      |                                        data_type:
[L:  5, P: 48]      |                                            datetime_type_identifier:
[L:  5, P: 48]      |                                                keyword:      'DATE'
[L:  5, P: 52]      |                                        [META] dedent:
[L:  5, P: 52]      |                                        end_bracket:          ')'
```

You can see the parse tree is correct.

**Comment by jmc-bbk** [comment]
Strange, rendering the template shows that the output file is as expected and that the fix operation works.

```
$ sqlfluff render <file>.sql > rendered.sql
$ sqlfluff fix rendered.sql

-- rendered.sql

SELECT table_a.*
FROM schema.table_a
  INNER JOIN schema.table_b
    ON
      table_a.col_a = table_b.col_a
      AND table_b.col_b = CAST('2022-09-23' AS DATE)
```

**Comment by jmc-bbk** [comment]
I've got a bit stuck on this one, perhaps need one of the maintainers to help out here.

```
CONFIG = FluffConfig.from_root(".")
lint = Linter(config=CONFIG)

# lint_string on the raw rendered string, returns expected violations/fixes
lint.lint_string(lint.render_file('SELECT.sql', CONFIG).templated_file.templated_str, config=CONFIG, encoding='utf8')

# lint_parsed on the file directly causes the issue outlined above (this path triggered via lint_path)
lint.lint_parsed(lint.parse_rendered(lint.render_file('SELECT.sql', CONFIG)), rule_pack)
```

I traced both of these back to `lint_fix_parsed` method and ruled out:

1/ Config is not an issue, there is a different memory instance passed in the above two approaches, but I hardcoded the config (to same instance) and confirmed it is **not** the root cause.

2/ Rules are not an issue, again a different memory instance was passed, but hardcoding the rules (to same instance) still produced the different behaviours from lint_string and lint_parsed.

---

`tree.raw` / `tree.stringify()` both return the same Tree

```
Tree Raw {'file': {'statement': {'select_statement': {'select_clause': {'keyword': 'SELECT', 'whitespace': ' ', 'select_clause_element': {'wildcard_expression': {'wildcard_identifier': {'naked_identifier': 'table_a', 'dot': '.', 'star': '*'}}}}, 'newline': '\n', 'from_clause': {'keyword': 'FROM', 'whitespace': ' ', 'from_expression': {'from_expression_element': {'table_expression': {'table_reference': [{'naked_identifier': 'schema'}, {'dot': '.'}, {'naked_identifier': 'table_a'}]}}, 'newline': '\n', 'whitespace': '  ', 'join_clause': [{'keyword': 'INNER'}, {'whitespace': ' '}, {'keyword': 'JOIN'}, {'whitespace': ' '}, {'from_expression_element': {'table_expression': {'table_reference': [{'naked_identifier': 'schema'}, {'dot': '.'}, {'naked_identifier': 'table_b'}]}}}, {'newline': '\n'}, {'whitespace': '    '}, {'join_on_condition': {'keyword': 'ON', 'whitespace': ' ', 'expression': [{'column_reference': [{'naked_identifier': 'table_b'}, {'dot': '.'}, {'naked_identifier': 'col_a'}]}, {'whitespace': ' '}, {'comparison_operator': {'raw_comparison_operator': '='}}, {'whitespace': ' '}, {'column_reference': [{'naked_identifier': 'table_a'}, {'dot': '.'}, {'naked_identifier': 'col_a'}]}, {'newline': '\n'}, {'whitespace': '      '}, {'binary_operator': 'AND'}, {'whitespace': ' '}, {'column_reference': [{'naked_identifier': 'table_b'}, {'dot': '.'}, {'naked_identifier': 'col_b'}]}, {'whitespace': ' '}, {'comparison_operator': {'raw_comparison_operator': '='}}, {'whitespace': ' '}, {'function': {'function_name': {'function_name_identifier': 'CAST'}, 'bracketed': [{'start_bracket': '('}, {'expression': {'quoted_literal': "'2022-09-23'"}}, {'whitespace': ' '}, {'keyword': 'AS'}, {'whitespace': ' '}, {'data_type': {'datetime_type_identifier': {'keyword': 'DATE'}}}, {'end_bracket': ')'}]}}]}}]}}}}, 'newline': '\n'}}
```

but checking out the attributes `vars(tree)` does show some different underlying data

```
# Tree Vars (string)
Tree Vars {'_file_path': '<string input>', 'pos_marker': PositionMarker(source_slice=slice(0, 155, None), templated_slice=slice(0, 155, None), templated_file=<TemplatedFile>, working_line_no=1, working_line_pos=1), 'segments': (<StatementSegment: ([L:  1, P:  1])>, <NewlineSegment: ([L:  5, P: 53]) '\n'>, <EndOfFile: ([L:  6, P:  1]) ''>), 'uuid': 196126503950348759087414225514274967820, 'raw': "SELECT table_a.*\nFROM schema.table_a\n  INNER JOIN schema.table_b\n    ON table_b.col_a = table_a.col_a\n      AND table_b.col_b = CAST('2022-09-23' AS DATE)\n"}

# Tree Vars (parsed)
Tree Vars {'_file_path': 'SELECT.sql', 'pos_marker': PositionMarker(source_slice=slice(0, 150, None), templated_slice=slice(0, 155, None), templated_file=<TemplatedFile>, working_line_no=1, working_line_pos=1), 'segments': (<StatementSegment: ([L:  1, P:  1])>, <NewlineSegment: ([L:  5, P: 48]) '\n'>, <EndOfFile: ([L:  6, P:  1]) ''>), 'uuid': 253065598784655715563041043697082033653, 'raw': "SELECT table_a.*\nFROM schema.table_a\n  INNER JOIN schema.table_b\n    ON table_b.col_a = table_a.col_a\n      AND table_b.col_b = CAST('2022-09-23' AS DATE)\n"}
```

**Comment by keraion** [comment]
This appears to be caused by the combination of LT02 and ST09 being applied to the same slice location. The fix skips over the overlapping slices dropping some of the input from the fixed output.

```
DEBUG      Filtered source patches:
DEBUG          0: FixPatch(templated_slice=slice(71, 72, None), fixed_raw='\n', patch_category='literal', source_slice=slice(71, 72, None), templated_str=' ', source_str=' ')
DEBUG          1: FixPatch(templated_slice=slice(72, 88, None), fixed_raw='      ', patch_category='mid_point', source_slice=slice(72, 88, None), templated_str='', source_str='')
DEBUG          2: FixPatch(templated_slice=slice(72, 85, None), fixed_raw='table_a.col_a', patch_category='literal', source_slice=slice(72, 85, None), templated_str='table_b.col_a', source_str='table_b.col_a')
DEBUG          3: FixPatch(templated_slice=slice(88, 101, None), fixed_raw='table_b.col_a', patch_category='literal', source_slice=slice(88, 101, None), templated_str='table_a.col_a', source_str='table_a.col_a')
DEBUG          4: FixPatch(templated_slice=slice(102, 106, None), fixed_raw='      ', patch_category='literal', source_slice=slice(102, 106, None), templated_str='    ', source_str='    ')
DEBUG          5: FixPatch(templated_slice=slice(152, 152, None), fixed_raw='\n', patch_category='end_point', source_slice=slice(147, 147, None), templated_str='', source_str='')
DEBUG      Source-only slices: []
INFO       Skipping overlapping patch at Index 88, Patch: FixPatch(templated_slice=slice(72, 85, None), fixed_raw='table_a.col_a', patch_category='literal', source_slice=slice(72, 85, None), templated_str='table_b.col_a', source_str='table_b.col_a')
```

https://github.com/sqlfluff/sqlfluff/blob/4572998ddc0dbdb3cd6139ef2b37387fb9993f13/src/sqlfluff/core/linter/linted_file.py#L308C1-L319C1

The most straightforward fix might be removing the `continue` in `LintedFile._slice_source_file_using_patches` and allowing all patches to be applied.

**Comment by jmc-bbk** [comment]
> This appears to be caused by the combination of LT02 and ST09 being applied to the same slice location. The fix skips over the overlapping slices dropping some of the input from the fixed output.
> 
> ```
> DEBUG      Filtered source patches:
> DEBUG          0: FixPatch(templated_slice=slice(71, 72, None), fixed_raw='\n', patch_category='literal', source_slice=slice(71, 72, None), templated_str=' ', source_str=' ')
> DEBUG          1: FixPatch(templated_slice=slice(72, 88, None), fixed_raw='      ', patch_category='mid_point', source_slice=slice(72, 88, None), templated_str='', source_str='')
> DEBUG          2: FixPatch(templated_slice=slice(72, 85, None), fixed_raw='table_a.col_a', patch_category='literal', source_slice=slice(72, 85, None), templated_str='table_b.col_a', source_str='table_b.col_a')
> DEBUG          3: FixPatch(templated_slice=slice(88, 101, None), fixed_raw='table_b.col_a', patch_category='literal', source_slice=slice(88, 101, None), templated_str='table_a.col_a', source_str='table_a.col_a')
> DEBUG          4: FixPatch(templated_slice=slice(102, 106, None), fixed_raw='      ', patch_category='literal', source_slice=slice(102, 106, None), templated_str='    ', source_str='    ')
> DEBUG          5: FixPatch(templated_slice=slice(152, 152, None), fixed_raw='\n', patch_category='end_point', source_slice=slice(147, 147, None), templated_str='', source_str='')
> DEBUG      Source-only slices: []
> INFO       Skipping overlapping patch at Index 88, Patch: FixPatch(templated_slice=slice(72, 85, None), fixed_raw='table_a.col_a', patch_category='literal', source_slice=slice(72, 85, None), templated_str='table_b.col_a', source_str='table_b.col_a')
> ```
> 
> https://github.com/sqlfluff/sqlfluff/blob/4572998ddc0dbdb3cd6139ef2b37387fb9993f13/src/sqlfluff/core/linter/linted_file.py#L308C1-L319C1
> 
> The most straightforward fix might be removing the `continue` in `LintedFile._slice_source_file_using_patches` and allowing all patches to be applied.

Thanks for checking this out @keraion, can I ask how you returned the above output with `DEBUG`? It's been a while since I've contributed, so I'm unsure on the best way to investigate these issues.

**Comment by keraion** [comment]
@jmc-bbk You can run any of the sqlfluff commands with `-vvvv` to get the debug level of logging.

My typical workflow:
```sh
mkdir test_sql/sqlfluff_5641
# add example query to this file:
test_sql/sqlfluff_5641/example.sql
# add config to this file:
test_sql/sqlfluff_5641/.sqlfluff
# make the templates directory
mkdir -p test_sql/sqlfluff_5641/noom_data_models/templates/
# run fix with debug verbose level
sqlfluff fix -vvvv --dialect redshift test_sql/sqlfluff_5641/example.sql
```
